proposals/2020-07-28_keyring-cmm/proposal.md

```diff
@@ -0,0 +1,209 @@
+we should define a CMM whose sole purpose is to be this base implementation.
+We will call this CMM implementation the Keyring CMM.
+For implementations which need to support master key provider, we will also define a
+Master Key Provider CMM.
+
+Similar to how we define a set of algorithm suites and define a default algorithm suite,
+we should similarly have a set of provided CMM implementations
+and a Default CMM which represents a particular configuration of one of those CMMs
+(possibly composed with other CMMs).
+We will define that the Default CMM is "the Keyring CMM".
+Just as an example, if we wanted to provide caching as a default behavior,
+the Default CMM could be defined as "the Caching CMM configured with a Keyring CMM."
+By defining the default in this way, we have greater flexibility to update the default
+CMM that customers use in the future without losing the composability of these implementations.
+
+Such a default MUST NOT allow any options other than the keyring to be configured in order
+to allow us to more easily maintain backwards compatability with possible Default CMM updates.
+
+## Drawbacks
+
+This change SHOULD NOT have any drawbacks.
+
+## Security Implications
+
+This change SHOULD NOT have any security implications.
+
+## Operational Implications
+
+This change SHOULD NOT have any operational implications.
+
+This change MUST maintain behavior for customers who are using a Default CMM.
+
+## Guide-level Explanation
+
+This change defines the [Keyring CMM](#keyring-cmm) and
+the [Master Key Provider CMM](#master-key-provider-cmm)
+as base CMM implementations that can be used "as is" or composed with other CMM.
+
+This change also defines a [Default CMM](#default-cmm) as a specific configuration of a CMM
+that the AWS Encryption SDK provides as a safe default.
+
+Customers SHOULD either use the [Default CMM](#default-cmm)
+or compose a CMM using the AWS Encryption SDK's provided CMM implementations
+(including the Keyring CMM).
+
+### Default CMM
+
+The Default CMM is a specific configuration of a
+CMM implementation provided by the AWS Encryption SDK.
+
+The specific CMM configuration describes a
+[safe default](https://github.com/awslabs/aws-encryption-sdk-specification/blob/2ec7674c304c408c2a32d412e834939c73f68d80/tenets.md#sensible-defaults)
+that serves most use cases.
+
+The CMM configuration defined by the Default CMM is the [Keyring CMM](#keyring-cmm) as is
+(The Keyring CMM provides no additional options other than specifying an underlying Keyring).
+
+As the AWS Encryption SDK provides more CMM implementations that would benefit default use cases,
+we expect to update the Default CMM to a configuration which composes such CMMs together provide
+useful properties.
+Any update to the composition defined by Default CMM SHOULD be backwards compatable.
+
+To ensure that its use is correct by construction,
+the instantiation of the Default CMM MUST NOT take any options
```

Comment:
```suggestion
the instantiation of the Default CMM MUST NOT take any options,
```

Reply:
I prefer without a comma here, but feel free to tell me I'm wrong.